### PR TITLE
Introduce the component dependency resolution entry point. 

### PR DESCRIPTION
**Pull request by vitek-karas**
- Refactors argument init to use the same code for app and components
- Remove usage of the global init structure from args init - makes it much clearer what is used where.
- Other small refactorings
- Adds basic doc with notes and open questions - should eventually become the real design doc for this feature.

### CODE REVIEW

**Comment by vitek-karas** [comment]
Still work in progress:
- Some TODOs left in the code - mostly to make sure we didn't forget about anything
- I added a doc describing the feature a little bit - not sure if we should keep it though

Note that for now I don't plan to fix the error logging and tracing in this PR. The error logging is a larger feature which we will need for other reasons and thus will be done separately later on.

File: Documentation/design-docs/host-component-dependencies-resolution.md [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+* Add ability to corelate tracing with the runtime - probably some kind of activity ID
+* Handling of native assets - currently returning just probing paths. Would be cleaner to return full resolved paths. But we would have to keep some probing paths. In the case of missing `.deps.json` the native library should be looked for in the component directory - thus requires probing - we can't figure out which of the files in the folder are native libraries in the hosts.
+* Handling of satellite assemblies (resource assets) - currently returning just probing paths which exclude the culture. So from a resolved asset `./foo/en-us/resource.dll` we only take `./foo` as the probing path. Consider using full paths instead - probably would require more parsing as we would have to be able to figure out the culture ID somewhere to build the true map AssemblyName->path in the managed class.
```
Comment (jeffschwMSFT) [inline Documentation/design-docs/host-component-dependencies-resolution.md]: The one open question I would like see added is around what sets of dependencies are returned by this API - just the app specific or the framework as well.  The comparison I think we should do is "does this feature completely remove the need for DependencyModel" or said another way "can DependencyModel now to be thin wrapper over this API?"  The killer scenario for me is a non-compiler tool trying to use the new version of introspective only load.

File: Documentation/design-docs/host-component-dependencies-resolution.md [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+  * probe_paths - reused
+  * tfm - reused
+  * host_mode - reused
+  * RID fallback graph from the root framework - reused
+* Just like settings, there's a set of environment variables which are used in the same way as the app would use them
+  * `DOTNET_ROLL_FORWARD_ON_NO_CANDIDATE_FX` - used just like in the app
+  * `ProgramFiles` and `ProgramFiles(x86")` - used to find servicing root and the shared store
+  * `DOTNET_SHARED_STORE` - used to find the shared store - just like the app
+  * `DOTNET_MULTILEVEL_LOOKUP` - used to enable multi-level lookup - used just like in the app
+* Right the capability doesn't process `.runtimeconfig.json` or `.runtimeconfig.dev.json`. Most dynamically loaded components don't have these anyway, since SDK doesn't generate these for the `classlib` project type. The only meaningful piece of info from these which could be used is the set of probing paths to use. Currently we reuse the same set as the app. With the changes in .NET Core 3 where `dotnet build` will copy all static dependencies locally, the importance of additional probing paths should be very low anyway.
```
Comment (swaroop-sridhar) [inline Documentation/design-docs/host-component-dependencies-resolution.md]: Typo: Right (now) ?

File: src/corehost/cli/deps_resolver.cpp [hunk 3]
```diff
@@ -597,7 +598,7 @@ void deps_resolver_t::resolve_additional_deps(const hostpolicy_init_t& init)
         return;
     }
 
-    pal::string_t additional_deps_serialized = init.additional_deps_serialized;
+    pal::string_t additional_deps_serialized = args.additional_deps_serialized;
```
Comment (swaroop-sridhar) [inline src/corehost/cli/deps_resolver.cpp]: Should additional-deps affect just the main app, or all plugins?

Comment (vitek-karas) [reply src/corehost/cli/deps_resolver.cpp]: For component dependency resolution the `args.additional_deps_serialized` will always be an empty string. Look at hostpolicy.cpp line 508 where the args are initialized.

Comment (vitek-karas) [reply src/corehost/cli/deps_resolver.cpp]: I added a test case for this to make sure.

File: Documentation/design-docs/host-component-dependencies-resolution.md (hunk 1 above)
Comment (vitek-karas) [reply Documentation/design-docs/host-component-dependencies-resolution.md]: I added new sections describing the Framework resolution and the relationship to DependencyModel. Can you please review them if they cover it?

**Comment by vitek-karas** [comment]
@sdmaclea There's no check to verify compatibility with the runtime or the frameworks of the app. This was an intentional choice. For most `classlibs` such checks are really tricky as they don't have `.runtimeconfig.json` so no explicit framework/runtime dependency is declared. In theory some parts of the `.deps.json` could be used for this, but it would require the code to hardcode well known framework names. Also components which are compiled as portable based on NETStandard don't have any framework dependencies as such. In such case the code would have to hardcode dependencies between portable packages and .NET core frameworks.

Another consideration is potential "roll-backwards" behavior. Component which was compiled against NETCore.App 2.1.0 may be loaded into an app using NETCore.App 2.0.0. If the component doesn't have references to assemblies of higher versions which are only available in 2.1.0, it would probably run fine against 2.0.0. We should leave such decisions to the runtime which can correctly determine assembly versions at load time, instead of broadly preventing such scenario from working.

**Comment by sdmaclea** [comment]
> no explicit framework/runtime dependency is declared.

We could fix this...

> Another consideration is potential "roll-backwards" behavior....

Roll backwards feels inherently risky.  Guidance should definitely not recommend this.  The big issue is late JIT exceptions as pieces of the required API are missing.  I think we will version the Assembly for the .NET Core framework uniformly which will nicely prevent rollback.  For other frameworks probably a similar thing is true.

I think I am OK with your proposal.

**Pull request merged**